### PR TITLE
Use LogSensitiveData flag to limit instance data in exception and log messages

### PR DESCRIPTION
**Pull request by mikary**
Fix for exception messages identified by #2752.
Promote `ISensitiveDataLogger` to core.

### CODE REVIEW

File: src/EntityFramework.Relational/Storage/RelationalConnection.cs [hunk 1]
```diff
@@ -217,8 +222,13 @@ public virtual void Open()
             {
                 _logger.LogVerbose(
                     RelationalLoggingEventId.OpeningConnection,
-                    ConnectionString,
-                    RelationalStrings.RelationalLoggerOpeningConnection);
+                    new RelationalConnectionLogData(
```
Comment (lajones) [inline src/EntityFramework.Relational/Storage/RelationalConnection.cs]: Nit. This seems like an odd way to do it, but if you can't pass `_connection.Value.Database` and `_connection.Value.DataSource` directly then OK.

File: src/EntityFramework.Core/Query/ExpressionVisitors/Internal/ParameterExtractingExpressionVisitor.cs [hunk 2]
```diff
@@ -18,25 +19,30 @@ public class ParameterExtractingExpressionVisitor : ExpressionVisitorBase
         public static Expression ExtractParameters(
             [NotNull] Expression expression,
             [NotNull] QueryContext queryContext,
-            [NotNull] IEvaluatableExpressionFilter evaluatableExpressionFilter)
+            [NotNull] IEvaluatableExpressionFilter evaluatableExpressionFilter,
+            [NotNull] ISensitiveDataLogger logger)
```
Comment (lajones) [inline src/EntityFramework.Core/Query/ExpressionVisitors/Internal/ParameterExtractingExpressionVisitor.cs]: Re the [NotNull], what gets passed here if `SensitiveDataLoggingEnabled` is false?

Comment (mikary) [inline src/EntityFramework.Core/Query/ExpressionVisitors/Internal/ParameterExtractingExpressionVisitor.cs]: `ISensitiveDataLogger` will exist either way, the flag is just a setting from `DbContextOptions` that can be accessed through this interface. Checking the value also logs a warning if sensitive data logging is enabled.

File: src/EntityFramework.Relational/Storage/RelationalConnection.cs (hunk 1 above)
Comment (mikary) [inline src/EntityFramework.Relational/Storage/RelationalConnection.cs]: Could work with an anonymous type. We use a strong type for LogCommand (`DbCommandLogData`) which is a public type. Not sure if there is a preference / pattern we want to establish for these.

Comment (anpete) [inline src/EntityFramework.Relational/Storage/RelationalConnection.cs]: Use a nominal type when it needs to be consumable. E.g. from tests.

File: src/EntityFramework.Relational/Storage/RelationalConnection.cs [hunk 3]
```diff
@@ -270,5 +285,19 @@ public virtual void Dispose()
                 _openedCount = 0;
             }
         }
+
+        private class RelationalConnectionLogData
```
Comment (anpete) [inline src/EntityFramework.Relational/Storage/RelationalConnection.cs]: This is either public or use an anon type.

File: src/EntityFramework.Core/Query/ExpressionVisitors/Internal/ParameterExtractingExpressionVisitor.cs (hunk 2 above)
Comment (lajones) [inline src/EntityFramework.Core/Query/ExpressionVisitors/Internal/ParameterExtractingExpressionVisitor.cs]: OK. No problem then.

**Comment by lajones** [comment]
LGTM :shipit:

**Pull request merged**